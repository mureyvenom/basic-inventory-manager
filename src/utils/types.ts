import { ReactNode } from 'react';

export interface InventoryItem {
  name: string;
  stock: number;
  price: number;
  description: string;
  owner: string;
  date_created: string;
  last_modified: string;
}

export interface UserContextType {
  user: {
    email: string;
  };
  setUser: (e: string) => void;
}

export interface InventoryContextType {
  items: InventoryItem[] | null;
  saveItems: (items: InventoryItem[]) => void;
  addItem: (item: InventoryItem) => boolean;
  updateItem: (original: InventoryItem, replace: InventoryItem) => boolean;
  deleteItem: (item: InventoryItem) => void;
  getItems: () => InventoryItem[];
}

export interface SettingsContextType {
  darkmode: boolean;
  toggleDarkMode: () => void;
}

export interface ProviderProps {
  children: ReactNode;
}
