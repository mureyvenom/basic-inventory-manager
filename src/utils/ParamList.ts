import { InventoryItem } from './types';

export type MainStack = {
  Signin: undefined;
  InventoryList: undefined;
  Inventory: {
    mode: 'add' | 'edit';
    inventory?: InventoryItem;
  };
};
