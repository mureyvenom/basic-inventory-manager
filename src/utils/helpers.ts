export const numberWithCommas = (x: number, decimals?: number) => {
  return x.toFixed(decimals || 0).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const addAlpha = (color: string, opacity: number) => {
  // coerce values so ti is between 0 and 1.
  var _opacity = Math.round(Math.min(Math.max(opacity || 1, 0), 1) * 255);
  return color + _opacity.toString(16).toUpperCase();
};

export const appConstants = () => {
  return {
    inventoryItemsKey: 'inventory-app-items',
    inventoryUserKey: 'inventory-app-user',
    inventorySettingsKey: 'inventory-app-settings',
  };
};

export const mockInventoryItems = [
  {
    date_created: '2023-03-11T02:27:30.776Z',
    description: 'Testing this out',
    last_modified: '2023-03-11T02:46:15.243Z',
    name: 'Testing this out',
    owner: 'holuwamurewa@gmail.com',
    price: 250,
    stock: 10,
  },
  {
    date_created: '2023-03-11T02:42:41.800Z',
    description: 'Second test',
    last_modified: '2023-03-11T02:48:05.826Z',
    name: 'Second test',
    owner: 'holuwamurewa@gmail.com',
    price: 1200,
    stock: 50,
  },
  {
    date_created: '2023-03-11T02:45:20.812Z',
    description: 'description',
    last_modified: '2023-03-11T02:45:20.812Z',
    name: 'Fourth entry',
    owner: 'holuwamurewa@gmail.com',
    price: 780,
    stock: 1000,
  },
];
