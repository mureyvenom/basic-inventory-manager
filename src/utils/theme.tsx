import { createTheme } from '@shopify/restyle';

const palette = {
  blue: '#6294CD',
  green: '#62CDA6',
  red: '#F92A4C',
  lightgreen: 'rgba(73,217,140,0.25)',
  lightred: 'rgba(255,102,127,0.25)',
  lightBg: 'rgba(255,255,255,0.05)',
  gray: '#2A3B3B',
  darkred: '#DA5757',
  darkgray: '#707070',
  black_: '#0A0A0A',
  lightBlue: '#29B9DC',
  purple: '#ED62F6',
  white: '#DDDDDD',
  muted: '#B0B0B0',
  faint: '#151515',
  transparent: 'transparent',
  constant_white: '#ffffff',
  constant_black: '#000000',
  // light theme colors
  bgLight: '#ffffff',
  faintLight: '#F9F9F9',
};

export const theme = createTheme({
  colors: {
    background: palette.black_,
    foreground: palette.white,
    primary: palette.blue,
    success: palette.green,
    danger: palette.red,
    lightsuccess: palette.lightgreen,
    lightdanger: palette.lightred,
    dangerdark: palette.darkred,
    secondary: palette.gray,
    info: palette.lightBlue,
    light: palette.purple,
    muted: palette.muted,
    muteddark: palette.darkgray,
    faint: palette.faint,
    white: palette.constant_white,
    black: palette.constant_black,
  },
  spacing: {
    xs: 4,
    s: 8,
    m: 16,
    l: 24,
    xl: 40,
    xxl: 80,
  },
  textVariants: {
    bold: {
      color: 'foreground',
      fontWeight: '700',
    },
    medium: {
      color: 'foreground',
      fontWeight: '600',
    },
    regular: {
      color: 'foreground',
    },
  },
  breakpoints: {},
});

export type Theme = typeof theme;

export const lightTheme: Theme = {
  ...theme,
  colors: {
    ...theme.colors,
    background: palette.bgLight,
    faint: palette.faintLight,
    foreground: palette.black_,
  },
};

export default theme;
