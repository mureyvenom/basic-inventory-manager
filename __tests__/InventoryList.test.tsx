import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { ThemeProvider } from '@shopify/restyle';
import theme from '../src/utils/theme';
import { render, fireEvent, act, screen } from '@testing-library/react-native';
import { MainStack } from '../src/utils/ParamList';
// import RootProvider from '../src/state/root.provider';
import InventoryList from '../src/screens/InventoryListScreen';
import { SettingsContext } from '../src/state/providers/settings.provider';
import { UserContext } from '../src/state/providers/user.provider';
import { InventoryContext } from '../src/state/providers/inventory.provider';
import { ProviderProps } from '../src/utils/types';
import { mockInventoryItems } from '../src/utils/helpers';
import { NavigationContainer } from '@react-navigation/native';
import MainStackNavigator from '../src/navigators/MainStack';

// jest.useFakeTimers();
// afterEach(cleanup);

let navigation: NativeStackNavigationProp<MainStack, 'InventoryList'>;

describe('Testing the inventory list screen', () => {
  const RootProvider = ({ children }: ProviderProps) => {
    const darkmode = true;
    const toggleDarkMode = jest.fn();
    const addItem = jest.fn();
    const deleteItem = jest.fn();
    const getItems = () => mockInventoryItems;
    const saveItems = jest.fn();
    const updateItem = jest.fn();
    const setUser = (_: string) => {};

    return (
      <SettingsContext.Provider
        value={{
          darkmode,
          toggleDarkMode,
        }}>
        <UserContext.Provider
          value={{
            setUser,
            user: {
              email: 'holuwamurewa@gmail.com',
            },
          }}>
          <InventoryContext.Provider
            value={{
              addItem,
              deleteItem,
              getItems,
              saveItems,
              updateItem,
              items: mockInventoryItems,
            }}>
            {children}
          </InventoryContext.Provider>
        </UserContext.Provider>
      </SettingsContext.Provider>
    );
  };

  it('Create a snapshot of the screen', async () => {
    const tree = renderer
      .create(
        <RootProvider>
          <ThemeProvider theme={theme}>
            <InventoryList navigation={navigation} />
          </ThemeProvider>
        </RootProvider>,
      )
      .toJSON();
    await renderer.act(async () => expect(tree).toMatchSnapshot());
  });

  it('Renders inventory items and navigates to inventory screen when item pressed', async () => {
    const renderedComponent = render(
      <RootProvider>
        <ThemeProvider theme={theme}>
          <NavigationContainer>
            <MainStackNavigator />
          </NavigationContainer>
        </ThemeProvider>
      </RootProvider>,
    );

    const inventoryItems = renderedComponent.getAllByTestId('inventory-item');
    expect(inventoryItems.length).toBeGreaterThan(0);

    await act(async () => fireEvent.press(inventoryItems[0]));
    screen.getByTestId('Delete');
  });
});
