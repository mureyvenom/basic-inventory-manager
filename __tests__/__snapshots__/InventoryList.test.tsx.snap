// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the inventory list screen Create a snapshot of the screen 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "flex": 1,
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 24,
            "paddingHorizontal": 24,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "marginRight": 24,
            },
          ]
        }
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <View
            style={
              [
                {
                  "width": "100%",
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "backgroundColor": "#6294CD",
                      "borderRadius": 6,
                      "flexDirection": "row",
                      "height": 50,
                      "justifyContent": "center",
                      "paddingHorizontal": 24,
                    },
                    {},
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "#ffffff",
                        "fontSize": 16,
                        "fontWeight": "600",
                      },
                    ]
                  }
                >
                  Add Item
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          [
            {
              "alignItems": "center",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#DDDDDD",
              },
            ]
          }
        >
          Dark Mode
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "height": 31,
              "width": 51,
            }
          }
          value={true}
        />
      </View>
    </View>
    <View
      style={
        [
          {
            "marginTop": 8,
            "paddingHorizontal": 24,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#DDDDDD",
            },
          ]
        }
      >
        Items Count: 
        3
      </Text>
    </View>
    <View
      style={
        [
          {
            "flex": 1,
            "marginVertical": 16,
            "paddingHorizontal": 24,
          },
        ]
      }
    >
      <RCTScrollView>
        <View>
          <View
            style={
              [
                {
                  "marginBottom": 16,
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="inventory-item"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#B0B0B0",
                      "borderBottomWidth": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 16,
                    },
                    {
                      "backgroundColor": "#DDDDDD1A",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "flex": 1,
                        "marginRight": 24,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "marginBottom": 24,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                              "fontSize": 16,
                              "fontWeight": "700",
                              "marginBottom": 8,
                            },
                          ]
                        }
                      >
                        Second test
                      </Text>
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                            },
                          ]
                        }
                      >
                        Price: 
                        1,200
                      </Text>
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                            },
                          ]
                        }
                      >
                        Stock: 
                        50
                      </Text>
                    </View>
                    <Text
                      style={
                        [
                          {
                            "color": "#DDDDDD",
                            "fontSize": 10,
                          },
                        ]
                      }
                    >
                      Last Modified: 
                      Sat Mar 11 2023
                       @
                       
                      03:48:05 GMT+0100 (West Africa Standard Time)
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "marginBottom": 16,
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="inventory-item"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#B0B0B0",
                      "borderBottomWidth": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 16,
                    },
                    {
                      "backgroundColor": "#DDDDDD1A",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "flex": 1,
                        "marginRight": 24,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "marginBottom": 24,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                              "fontSize": 16,
                              "fontWeight": "700",
                              "marginBottom": 8,
                            },
                          ]
                        }
                      >
                        Testing this out
                      </Text>
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                            },
                          ]
                        }
                      >
                        Price: 
                        250
                      </Text>
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                            },
                          ]
                        }
                      >
                        Stock: 
                        10
                      </Text>
                    </View>
                    <Text
                      style={
                        [
                          {
                            "color": "#DDDDDD",
                            "fontSize": 10,
                          },
                        ]
                      }
                    >
                      Last Modified: 
                      Sat Mar 11 2023
                       @
                       
                      03:46:15 GMT+0100 (West Africa Standard Time)
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "marginBottom": 16,
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="inventory-item"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#B0B0B0",
                      "borderBottomWidth": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 16,
                    },
                    {
                      "backgroundColor": "#DDDDDD1A",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "flex": 1,
                        "marginRight": 24,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "marginBottom": 24,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                              "fontSize": 16,
                              "fontWeight": "700",
                              "marginBottom": 8,
                            },
                          ]
                        }
                      >
                        Fourth entry
                      </Text>
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                            },
                          ]
                        }
                      >
                        Price: 
                        780
                      </Text>
                      <Text
                        style={
                          [
                            {
                              "color": "#DDDDDD",
                            },
                          ]
                        }
                      >
                        Stock: 
                        1,000
                      </Text>
                    </View>
                    <Text
                      style={
                        [
                          {
                            "color": "#DDDDDD",
                            "fontSize": 10,
                          },
                        ]
                      }
                    >
                      Last Modified: 
                      Sat Mar 11 2023
                       @
                       
                      03:45:20 GMT+0100 (West Africa Standard Time)
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        [
          {
            "height": 90,
            "justifyContent": "center",
            "paddingHorizontal": 24,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "width": "100%",
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#F92A4C",
                  "borderRadius": 6,
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "center",
                  "paddingHorizontal": 24,
                },
                {},
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                    "fontSize": 16,
                    "fontWeight": "600",
                  },
                ]
              }
            >
              Sign out of holuwamurewa@gmail.com
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    testID="confirmation-modal"
    transparent={true}
    visible={false}
  />
</RCTSafeAreaView>
`;
